pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1322 @@
+		ValBondBelowMin,
+		ValBondAboveMax,
+		NomBondBelowMin,
+		DelegationBelowMin,
+		AlreadyOffline,
+		AlreadyActive,
+		AlreadyLeaving,
+		TooManyDelegators,
+		CannotActivateIfLeaving,
+		ExceedMaxCollatorsPerNom,
+		AlreadyDelegatedCollator,
+		DelegationDNE,
+		Underflow,
+		InvalidSchedule,
+		CannotSetBelowMin,
+		RewardLockDNE,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Starting Block, Round, Number of Collators Selected, Active Collator
+		/// Stake, Delegator Stake
+		NewRound(T::BlockNumber, RoundIndex, u32, BalanceOf<T>, BalanceOf<T>),
+		/// Account, Amount Locked by candidate, New Total Locked Collator
+		/// Candidate Amount
+		JoinedCollatorCandidates(T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		/// Round, Collator Account, Collator Self Bond, Delegator Stake (total)
+		CollatorChosen(RoundIndex, T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		/// Collator Account, Old Bond, New Bond
+		CollatorBondedMore(T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		/// Collator Account, Old Bond, New Bond
+		CollatorBondedLess(T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		CollatorWentOffline(RoundIndex, T::AccountId),
+		CollatorBackOnline(RoundIndex, T::AccountId),
+		/// Round, Collator Account, Scheduled Exit
+		CollatorScheduledExit(RoundIndex, T::AccountId, RoundIndex),
+		/// Account, Amount Unlocked, New Total Collator Amount Locked, New
+		/// Total Delegator Amount Locked
+		CollatorLeft(T::AccountId, BalanceOf<T>, BalanceOf<T>, BalanceOf<T>),
+		// Delegator, Collator, Old Delegation, New Delegation
+		DelegationIncreased(T::AccountId, T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		// Delegator, Collator, Old Delegation, New Delegation
+		DelegationDecreased(T::AccountId, T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		/// Delegator, Amount Unstaked
+		DelegatorLeft(T::AccountId, BalanceOf<T>),
+		/// Delegator, Amount Locked, Collator, New Total Amount backing
+		/// Collator
+		Delegation(T::AccountId, BalanceOf<T>, T::AccountId, BalanceOf<T>),
+		/// Delegator, Collator, Amount Unstaked, New Total Amount Staked for
+		/// Collator
+		DelegatorLeftCollator(T::AccountId, T::AccountId, BalanceOf<T>, BalanceOf<T>),
+		/// Paid the account (delegator or collator) the balance as liquid
+		/// rewards
+		Rewarded(T::AccountId, BalanceOf<T>),
+		/// Round inflation range set with the provided annual inflation range
+		RoundInflationSet(Perbill, Perbill, Perbill, Perbill),
+		/// Set total selected candidates to this value [old, new]
+		TotalSelectedSet(u32, u32),
+		/// Set blocks per round [current_round, first_block, old, new]
+		BlocksPerRoundSet(RoundIndex, T::BlockNumber, u32, u32),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(n: T::BlockNumber) {
+			let mut round = <Round<T>>::get();
+			if round.should_update(n) {
+				// kill snapshot of current round
+				<AtStake<T>>::remove_prefix(round.current);
```

Comment:
I don't know approximately how big this iteration will be, but you def can NOT do this on a on_finalize.

Should be moved to `on_initialize`, or better cleaned up via transactions.

Remember if for whatever reason this ends up too big, and it pushes the PoV time or computation time over the limit, you will have bricked your chain.

Reply:
Moved to `on_initialize` in 63fa2e4.

I don't know how I we could solve this in transactions: Both the collator exits as well as setting the new candidates should be applied automatically every new round.